editor/svgutils.js

```diff
@@ -529,19 +529,387 @@ svgedit.utilities.getBBox = function(elem) {
+	var d = '';
+
+	$.each(pathSegments, function(j, seg) {
+		var i;
+		var l = seg[0], pts = seg[1];
+		d += l;
+		for (i = 0; i < pts.length; i+=2) {
+			d += (pts[i] +','+pts[i+1]) + ' ';
+		}
+	});
+
+	return d
+}
+
+svgedit.utilities.getPathDFromElement = function( elem) {
+
+	// Possibly the cubed root of 6, but 1.81 works best
+	var num = 1.81;
+	var d, a, rx, ry;
+	switch (elem.tagName) {
+		case 'ellipse':
+		case 'circle':
+			a = $(elem).attr(['rx', 'ry', 'cx', 'cy']);
+			var cx = a.cx, cy = a.cy;
+			rx = a.rx;
+			ry = a.ry;
+			if (elem.tagName == 'circle') {
+				rx = ry = $(elem).attr('r');
+			}
+
+			d = svgedit.utilities.getPathDFromSegments([
+				['M',[(cx-rx),(cy)]],
+				['C',[(cx-rx),(cy-ry/num), (cx-rx/num),(cy-ry), (cx),(cy-ry)]],
+				['C',[(cx+rx/num),(cy-ry), (cx+rx),(cy-ry/num), (cx+rx),(cy)]],
+				['C',[(cx+rx),(cy+ry/num), (cx+rx/num),(cy+ry), (cx),(cy+ry)]],
+				['C',[(cx-rx/num),(cy+ry), (cx-rx),(cy+ry/num), (cx-rx),(cy)]],
+				['Z',[]]
+			]);
+			break;
+		case 'path':
+			d = elem.getAttribute('d');
+			break;
+		case 'line':
+			a = $(elem).attr(['x1', 'y1', 'x2', 'y2']);
+			d = 'M'+a.x1+','+a.y1+'L'+a.x2+','+a.y2;
+			break;
+		case 'polyline':
+			d = 'M' + elem.getAttribute('points');
+			break;
+		case 'polygon':
+			d = 'M' + elem.getAttribute('points') + ' Z';
+			break;
+		case 'rect':
+			var r = $(elem).attr(['rx', 'ry']);
+			rx = r.rx;
+			ry = r.ry;
+			var b = elem.getBBox();
+			var x = b.x, y = b.y, w = b.width, h = b.height;
+			num = 4 - num; // Why? Because!
+
+			if (!rx && !ry) {
+				// Regular rect
+				d = svgedit.utilities.getPathDFromSegments([
+					['M',[x, y]],
+					['L',[x+w, y]],
+					['L',[x+w, y+h]],
+					['L',[x, y+h]],
+					['L',[x, y]],
+					['Z',[]]
+				]);
+			} else {
+				d = svgedit.utilities.getPathDFromSegments([
+					['M',[x, y+ry]],
+					['C',[x, y+ry/num, x+rx/num, y, x+rx, y]],
+					['L',[x+w-rx, y]],
+					['C',[x+w-rx/num, y, x+w, y+ry/num, x+w, y+ry]],
+					['L',[x+w, y+h-ry]],
+					['C',[x+w, y+h-ry/num, x+w-rx/num, y+h, x+w-rx, y+h]],
+					['L',[x+rx, y+h]],
+					['C',[x+rx/num, y+h, x, y+h-ry/num, x, y+h-ry]],
+					['L',[x, y+ry]],
+					['Z',[]]
+				]);
+			}
+			break;
+		default:
+			break;
+	}
+
+	return d;
+
+};
+
+svgedit.utilities.addAttributesForConvertToPath = function( elem, attrs) {
+	// TODO: make this list global so that we can properly maintain it
+	// TODO: what about @transform, @clip-rule, @fill-rule, etc?
+	$.each(['marker-start', 'marker-end', 'marker-mid', 'filter', 'clip-path'], function() {
+		if (elem.getAttribute(this)) {
+			attrs[this] = elem.getAttribute(this);
+		}
+	});
+};
+
+// Function: getBBoxOfElementAsPath
+// Get the BBox of an element-as-path
 //
 // Parameters:
-// elem - Optional DOM element to get the angle for
+// elem - The DOM element to be probed
+// addSvgElementFromJson - Function to add the path element to the current layer. See canvas.addSvgElementFromJso
```

Comment:
I'll be fixing the spaces and semicolons; but I wanted to give feedback on this for now...

**Not adding to DOM:**
I've had issues in the past where, if the element is not attached to the DOM, the native getBBox returns zero width and height (in Chrome specifically). I'm also worried that it might work on some browser versions/platforms, but not others.

**Why create a path at all**
getBBoxWithTransform calls getBBoxOfElementAsPath when there is a simple shape with a transform. I assume it's doing this because (for example) a _rounded_ rect rotated 45 is not as wide as a _non-rounded_ rect rotated 45.

When it's not a simple shape, the new code gets the native bbox, then does matrix transforms on the bbox. This is because all the current browsers return the bbox without top-level transforms (and why the old code wrapped the transform inside a new group, then got the bbox).

Oddly, I tested the current release and the new code. The selections are the same before and after in Chrome and Safari. Firefox does seem to calculate a tighter bbox in the current released code (though with performance issues).

Here's an example with rotated ellipse and rotate g. Try rubber band selection starting from below to see if the bottom corner of the circle selection is part of the rubber band hit. The only advantage for the old code is with Firefox.
```
  <g id="svg_3" transform="rotate(45, 100, 100)">
   <ellipse fill="#56ff56" stroke-width="5" stroke-dasharray="null" stroke-linejoin="null" stroke-linecap="null" cx="100" cy="100" rx="50" ry="50" id="svg_1"/>
   <ellipse fill="#00bf5f" stroke-width="5" stroke-dasharray="null" stroke-linejoin="null" stroke-linecap="null" cx="116" cy="100" rx="20" ry="20" id="svg_2"/>
  </g>
  <ellipse transform="rotate(45, 300, 100)" ry="50" rx="50" id="svg_4" cy="100" cx="300" stroke-width="5" stroke="#000000" fill="#ff0000"/>
```

**Injected Functions**
In general, I like injected _services_ because it makes it easier to create unit tests.

I agree it would be much better if utility functions didn't depend on canvas. This is a refactoring of the original code in canvas.convertToPath. I broke it out into two functions, changing convertToPath internally, but not updating references _using_ convertToPath. I wanted to get feedback before taking it further.

**Next Steps**
Interestingly, the new code doesn't affect the selection handles bbox you see when you click on an object.

I think the slightly optimized bbox (in Firefox only) is not worth the performance hit.

We could remove the getBBoxOfElementAsPath calls in getBBoxWithTransform if we can live with the sub-optimal bbox hit-test for rubber band selection. Given this, we can remove both injected functions from the utilities.

**Future**
We could still optimize some shapes manually in the future. For example, it would be relatively easy to calculate the actual width of a rotated rounded rect. A more advanced transform would be trickier.

There is a getBoundingClientRect() that calculates the bbox with the transform which sounds like exactly what we need; however, it would have to be translated back to SVG space and I didn't have time to investigate that further.